routing/dht/dht_bootstrap.go

```diff
@@ -0,0 +1,185 @@
+func (dht *IpfsDHT) Bootstrap(ctx context.Context) (goprocess.Process, error) {
+
+	if err := dht.runBootstrap(ctx, DefaultBootstrapQueries); err != nil {
+		return nil, err
+	}
+
+	sig := time.Tick(DefaultBootstrapPeriod)
+	return dht.BootstrapOnSignal(DefaultBootstrapQueries, sig)
+}
+
+// SignalBootstrap ensures the dht routing table remains healthy as peers come and go.
+// it builds up a list of peers by requesting random peer IDs. The Bootstrap
+// process will run a number of queries each time, and run every time signal fires.
+// These parameters are configurable.
+//
+// SignalBootstrap returns a process, so the user can stop it.
+func (dht *IpfsDHT) BootstrapOnSignal(queries int, signal <-chan time.Time) (goprocess.Process, error) {
+	if queries <= 0 {
+		return nil, fmt.Errorf("invalid number of queries: %d", queries)
+	}
+
+	if signal == nil {
+		return nil, fmt.Errorf("invalid signal: %v", signal)
+	}
+
+	proc := periodicproc.Ticker(signal, func(worker goprocess.Process) {
+		// it would be useful to be able to send out signals of when we bootstrap, too...
+		// maybe this is a good case for whole module event pub/sub?
+
+		ctx := dht.Context()
+		if err := dht.runBootstrap(ctx, queries); err != nil {
+			log.Error(err)
+			// A bootstrapping error is important to notice but not fatal.
+			// maybe the client should be able to consume these errors,
+			// though I dont have a clear use case in mind-- what **could**
+			// the client do if one of the bootstrap calls fails?
+			//
+			// This is also related to the core's bootstrap failures.
+			// superviseConnections should perhaps allow clients to detect
+			// bootstrapping problems.
+			//
+			// Anyway, passing errors could be done with a bootstrapper object.
+			// this would imply the client should be able to consume a lot of
+			// other non-fatal dht errors too. providing this functionality
+			// should be done correctly DHT-wide.
+			// NB: whatever the design, clients must ensure they drain errors!
+			// This pattern is common to many things, perhaps long-running services
+			// should have something like an ErrStream that allows clients to consume
+			// periodic errors and take action. It should allow the user to also
+			// ignore all errors with something like an ErrStreamDiscard. We should
+			// study what other systems do for ideas.
+		}
+	})
+
+	return proc, nil
+}
+
+// runBootstrap builds up list of peers by requesting random peer IDs
+func (dht *IpfsDHT) runBootstrap(ctx context.Context, queries int) error {
+	bslog := func(msg string) {
+		log.Debugf("DHT %s dhtRunBootstrap %s -- routing table size: %d", dht.self, msg, dht.routingTable.Size())
+	}
+	bslog("start")
+	defer bslog("end")
+	defer log.EventBegin(ctx, "dhtRunBootstrap").Done()
+
+	var merr u.MultiErr
+
+	randomID := func() peer.ID {
+		// 16 random bytes is not a valid peer id. it may be fine becuase
+		// the dht will rehash to its own keyspace anyway.
```

Comment:
Huh, i could have sworn that I called `u.Hash` on the random bytes we read here.